tensorflow_lite_support/python/test/task/vision/object_detector_test.py

```diff
@@ -0,0 +1,118 @@
+from tensorflow_lite_support.python.task.processor.proto import class_pb2
+from tensorflow_lite_support.python.task.processor.proto import bounding_box_pb2
+from tensorflow_lite_support.python.task.processor.proto import detection_options_pb2
+from tensorflow_lite_support.python.task.processor.proto import detections_pb2
+from tensorflow_lite_support.python.task.vision import object_detector
+from tensorflow_lite_support.python.task.vision.core import tensor_image
+from tensorflow_lite_support.python.test import base_test
+from tensorflow_lite_support.python.test import test_util
+
+_BaseOptions = task_options.BaseOptions
+_ExternalFile = task_options.ExternalFile
+_ObjectDetector = object_detector.ObjectDetector
+_ObjectDetectorOptions = object_detector.ObjectDetectorOptions
+
+_MODEL_FILE = 'coco_ssd_mobilenet_v1_1.0_quant_2018_06_29.tflite'
+_IMAGE_FILE = 'cats_and_dogs.jpg'
+_ACCEPTABLE_ERROR_RANGE = 0.000001
+
+
+class ModelFileType(enum.Enum):
+  FILE_CONTENT = 1
+  FILE_NAME = 2
+
+
+class ObjectDetectorTest(parameterized.TestCase, base_test.BaseTestCase):
+
+  def setUp(self):
+    super().setUp()
+    self.test_image_path = test_util.get_test_data_path(_IMAGE_FILE)
+    self.model_path = test_util.get_test_data_path(_MODEL_FILE)
+
+  @staticmethod
+  def create_detector_from_options(model_file, **detection_options):
+    base_options = _BaseOptions(model_file=model_file)
+    detection_options = detection_options_pb2.DetectionOptions(
+        **detection_options)
+    options = _ObjectDetectorOptions(
+        base_options=base_options,
+        detection_options=detection_options)
+    detector = _ObjectDetector.create_from_options(options)
+    return detector
+
+  @staticmethod
+  def build_test_data(expected_detections):
+    expected_result = detections_pb2.DetectionResult()
+
+    for index in range(len(expected_detections)):
+      bounding_box, category = expected_detections[index]
+      detection = detections_pb2.Detection()
+      detection.bounding_box.CopyFrom(
+        bounding_box_pb2.BoundingBox(**bounding_box))
+      detection.classes.append(class_pb2.Category(**category))
+      expected_result.detections.append(detection)
+
+    expected_result_dict = json.loads(
+      json_format.MessageToJson(expected_result))
+
+    return expected_result_dict
+
+  def test_top1_result(self):
+    # Creates detector.
+    model_file = _ExternalFile(file_name=self.model_path)
+
+    detector = self.create_detector_from_options(
+      model_file=model_file, max_results=1)
+
+    # Loads image.
+    image = tensor_image.TensorImage.from_file(self.test_image_path)
+
+    # Performs object detection on the input.
+    image_result = detector.detect(image)
+    image_result_dict = json.loads(json_format.MessageToJson(image_result))
+
+    # Builds test data.
+    expected_detections = [
+      (
+        {
+          'origin_x': 54, 'origin_y': 396, 'width': 393, 'height': 196
```

Comment:
Hi.  Firstly, thank you for your contributions!

Secondly, I have some feedback on this PR.
This is extremely brittle and fragile test code.
Please don't write tests like this in future.
TensorFlow Lite does not guarantee bit-exact results.
The output from an ML model may vary depending on the details of exactly how (and in what order) the intermediate floating point calculations were done.

Rather than testing for exact values for the bounding box coordinates,
it is better to just verify that the computed bounding box coordinates are
within a reasonable range (e.g. a few pixels) of the expected bounding box.

The code in this CL does use an _ACCEPTABLE_ERROR_RANGE = 0.000001 parameter
to do the comparison, but that error margin is much too small for comparing pixel
coordinate values -- an error of 1 part in a million is much less than one pixel.

Reply:
Hi @fergushenderson. At the time of implementing the tests for these APIs I was merely following the C++ API and was expected to keep it consistent with those APIs. Those APIs happened to test the exact values within a reasonable range and I did too. So there’s that. Hope it clarifies everything.

Reply:
https://github.com/tensorflow/tflite-support/blob/master/tensorflow_lite_support/cc/test/task/vision/object_detector_test.cc

For your reference.

Reply:
https://github.com/tensorflow/tflite-support/blob/fe9f90bf2a3fd15352b1ed1399937b1fc859c235/tensorflow_lite_support/cc/test/task/vision/object_detector_test.cc#L113

Mostly followed this at the time of development.

Reply:
I think all of the tests in object_detector_test.cc that test for exact bounding box values are using `PostprocessTest`.  Those tests aren't getting the bounding box data from TensorFlow Lite, instead they mock out the object detection code, supplying hard-coded values for the bounding box location data,

https://github.com/tensorflow/tflite-support/blob/fe9f90bf2a3fd15352b1ed1399937b1fc859c235/tensorflow_lite_support/cc/test/task/vision/object_detector_test.cc#L446

and are just testing the post-processing.  So in those tests, it is probably fine to test for exact bounding box values.

However, the fact that these tests mock out the object detection code is not immediately clear from the code of the tests, instead being far away in the FillAndGetOutputTensors method of the PostprocessTest class, so I can see why this distinction wasn't obvious! :)

Reply:
@fergushenderson As for the post processing test cases that holds true. However the C++ tests did indeed compare exact bounding box values after loading a real TF Lite model (with metadata) and a real image as you can see here.

https://github.com/tensorflow/tflite-support/blob/fe9f90bf2a3fd15352b1ed1399937b1fc859c235/tensorflow_lite_support/cc/test/task/vision/object_detector_test.cc#L327

This does use the same `ExpectApproximatelyEqual` to compare values FYI.
In that case, I believe the C++ tests need to be revised as well by using a margin for the pixel coordinates error to compare the actual and expected outputs.

Reply:
The post processing tests aren’t involved in those Detect tests.

Reply:
> the C++ tests did indeed compare exact bounding box values after loading a real TF Lite model

OK, good point, I stand corrected.

> In that case, I believe the C++ tests need to be revised as well by using a margin for the pixel coordinates error to compare the actual and expected output

Agreed.  It looks like Alan Kelly <https://github.com/alankelly> has made a change for that already, but for some reason it hasn't propagated into github master yet.